### PR TITLE
✨Destination Bigquery 1s1t: handle cursor change

### PR DESCRIPTION
**Pull request by edgao**
Handle this scenario:
1. Source declares a cursor `old_cursor`; we run a sync
2. Source changes its schema to add a `new_cursor` column, and changes its cursor to `new_cursor`; run a sync

During the second sync, all old records have a null `new_cursor` (because that field didn't exist previously). But we still want to be able to overwrite them in the second sync. Therefore, we need to treat NULL cursors as older than non-NULL cursors.

* Update the base T+D tests to verify the schema change case.
* Update bigquery sqlgenerator integration test to verify the NULL cursor handling
* Update bigquery sqlgenerator to sort `NULLS LAST`

This is essentially happening with CDC sources right now.

### CODE REVIEW

File: airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/typing_deduping/BigQuerySqlGenerator.java [hunk 1]
```diff
@@ -503,7 +503,7 @@ String dedupFinalTable(final StreamId id,
               `_airbyte_raw_id` IN (
                 SELECT `_airbyte_raw_id` FROM (
                   SELECT `_airbyte_raw_id`, row_number() OVER (
-                    PARTITION BY ${pk_list} ORDER BY ${cursor_name} DESC, `_airbyte_extracted_at` DESC
+                    PARTITION BY ${pk_list} ORDER BY ${cursor_name} DESC NULLS LAST, `_airbyte_extracted_at` DESC
```
Comment (evantahler) [inline airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/typing_deduping/BigQuerySqlGenerator.java:506]: * [Snowflake](https://docs.snowflake.com/en/sql-reference/constructs/order-by) can do this! 
* [Redshift](https://docs.aws.amazon.com/redshift/latest/dg/r_ORDER_BY_clause.html) can do this!
* [Postgres](https://www.postgresql.org/docs/current/queries-order.html#:~:text=The%20NULLS%20FIRST%20and%20NULLS,order%2C%20and%20NULLS%20LAST%20otherwise.) can do this!
* [Clickhouse](https://clickhouse.com/docs/en/sql-reference/statements/select/order-by) can do this!
* [MySQL](https://www.designcise.com/web/tutorial/how-to-order-null-values-first-or-last-in-mysql) can do this (in a very odd way)!
* [Databrikcs](https://docs.databricks.com/sql/language-manual/sql-ref-syntax-qry-select-orderby.html) can do this! 

Seems like this is a thing for enough databases to roll with!

**Comment by octavia-squidington-iii** [comment]
##  destination-bigquery test report (commit [`043b8f5ac9`](https://github.com/airbytehq/airbyte/pull/28721/commits/043b8f5ac915cc83b1ded856d93016afbdf1f1df)) - ✅

⏲️  Total pipeline duration: 40mn40s 

| Step                                                                            | Result   |
|:--------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-bigquery/metadata.yaml     | ✅        |
| Connector version semver check                                                  | ✅        |
| QA checks                                                                       | ✅        |
| Build connector tar                                                             | ✅        |
| Build destination-bigquery docker image for platform linux/x86_64               | ✅        |
| Build airbyte/normalization:dev                                                 | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-bigquery:integrationTest | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/edgao_dest_bigquery_null_order/1690327609/043b8f5ac915cc83b1ded856d93016afbdf1f1df/destination-bigquery/1.5.8/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery test
```

**Comment by octavia-squidington-iii** [comment]
##  destination-bigquery-denormalized test report (commit [`043b8f5ac9`](https://github.com/airbytehq/airbyte/pull/28721/commits/043b8f5ac915cc83b1ded856d93016afbdf1f1df)) - ✅

⏲️  Total pipeline duration: 22mn02s 

| Step                                                                                         | Result   |
|:---------------------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-bigquery-denormalized/metadata.yaml     | ✅        |
| Connector version semver check                                                               | ✅        |
| QA checks                                                                                    | ✅        |
| Build connector tar                                                                          | ✅        |
| Build destination-bigquery-denormalized docker image for platform linux/x86_64               | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-bigquery-denormalized:integrationTest | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/edgao_dest_bigquery_null_order/1690327609/043b8f5ac915cc83b1ded856d93016afbdf1f1df/destination-bigquery-denormalized/1.5.3/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery-denormalized test
```

**Comment by octavia-squidington-iii** [comment]
##  destination-snowflake test report (commit [`043b8f5ac9`](https://github.com/airbytehq/airbyte/pull/28721/commits/043b8f5ac915cc83b1ded856d93016afbdf1f1df)) - ✅

⏲️  Total pipeline duration: 59mn17s 

| Step                                                                             | Result   |
|:---------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-snowflake/metadata.yaml     | ✅        |
| Connector version semver check                                                   | ✅        |
| QA checks                                                                        | ✅        |
| Build connector tar                                                              | ✅        |
| Build destination-snowflake docker image for platform linux/x86_64               | ✅        |
| Build airbyte/normalization-snowflake:dev                                        | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-snowflake:integrationTest | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/edgao_dest_bigquery_null_order/1690327609/043b8f5ac915cc83b1ded856d93016afbdf1f1df/destination-snowflake/1.2.4/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-snowflake test
```

**Comment by octavia-squidington-iii** [comment]
##  destination-snowflake test report (commit [`dfd2b58423`](https://github.com/airbytehq/airbyte/pull/28721/commits/dfd2b584235de211f6fc30389cd6a1c0aa083ec0)) - ✅

⏲️  Total pipeline duration: 59mn47s 

| Step                                                                             | Result   |
|:---------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-snowflake/metadata.yaml     | ✅        |
| Connector version semver check                                                   | ✅        |
| QA checks                                                                        | ✅        |
| Build connector tar                                                              | ✅        |
| Build destination-snowflake docker image for platform linux/x86_64               | ✅        |
| Build airbyte/normalization-snowflake:dev                                        | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-snowflake:integrationTest | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/edgao_dest_bigquery_null_order/1690335078/dfd2b584235de211f6fc30389cd6a1c0aa083ec0/destination-snowflake/1.2.4/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-snowflake test
```

**Comment by octavia-squidington-iii** [comment]
##  destination-bigquery-denormalized test report (commit [`dfd2b58423`](https://github.com/airbytehq/airbyte/pull/28721/commits/dfd2b584235de211f6fc30389cd6a1c0aa083ec0)) - ✅

⏲️  Total pipeline duration: 22mn37s 

| Step                                                                                         | Result   |
|:---------------------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-bigquery-denormalized/metadata.yaml     | ✅        |
| Connector version semver check                                                               | ✅        |
| QA checks                                                                                    | ✅        |
| Build connector tar                                                                          | ✅        |
| Build destination-bigquery-denormalized docker image for platform linux/x86_64               | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-bigquery-denormalized:integrationTest | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/edgao_dest_bigquery_null_order/1690335078/dfd2b584235de211f6fc30389cd6a1c0aa083ec0/destination-bigquery-denormalized/1.5.3/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery-denormalized test
```

**Comment by octavia-squidington-iii** [comment]
##  destination-bigquery test report (commit [`dfd2b58423`](https://github.com/airbytehq/airbyte/pull/28721/commits/dfd2b584235de211f6fc30389cd6a1c0aa083ec0)) - ✅

⏲️  Total pipeline duration: 38mn13s 

| Step                                                                            | Result   |
|:--------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-bigquery/metadata.yaml     | ✅        |
| Connector version semver check                                                  | ✅        |
| QA checks                                                                       | ✅        |
| Build connector tar                                                             | ✅        |
| Build destination-bigquery docker image for platform linux/x86_64               | ✅        |
| Build airbyte/normalization:dev                                                 | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-bigquery:integrationTest | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/edgao_dest_bigquery_null_order/1690335078/dfd2b584235de211f6fc30389cd6a1c0aa083ec0/destination-bigquery/1.5.8/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery test
```

**Review by evantahler — APPROVED** [review]

**Comment by octavia-squidington-iii** [comment]
##  destination-bigquery test report (commit [`face39e39e`](https://github.com/airbytehq/airbyte/pull/28721/commits/face39e39e1e5e67ede58c620f4e1a85ce32cf25)) - ✅

⏲️  Total pipeline duration: 31mn50s 

| Step                                                                            | Result   |
|:--------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-bigquery/metadata.yaml     | ✅        |
| Connector version semver check                                                  | ✅        |
| QA checks                                                                       | ✅        |
| Build connector tar                                                             | ✅        |
| Build destination-bigquery docker image for platform linux/x86_64               | ✅        |
| Build airbyte/normalization:dev                                                 | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-bigquery:integrationTest | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/edgao_dest_bigquery_null_order/1690412428/face39e39e1e5e67ede58c620f4e1a85ce32cf25/destination-bigquery/1.5.8/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery test
```

**Comment by octavia-squidington-iii** [comment]
##  destination-snowflake test report (commit [`face39e39e`](https://github.com/airbytehq/airbyte/pull/28721/commits/face39e39e1e5e67ede58c620f4e1a85ce32cf25)) - ✅

⏲️  Total pipeline duration: 43mn40s 

| Step                                                                             | Result   |
|:---------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-snowflake/metadata.yaml     | ✅        |
| Connector version semver check                                                   | ✅        |
| QA checks                                                                        | ✅        |
| Build connector tar                                                              | ✅        |
| Build destination-snowflake docker image for platform linux/x86_64               | ✅        |
| Build airbyte/normalization-snowflake:dev                                        | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-snowflake:integrationTest | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/edgao_dest_bigquery_null_order/1690412428/face39e39e1e5e67ede58c620f4e1a85ce32cf25/destination-snowflake/1.2.4/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-snowflake test
```

**Comment by octavia-squidington-iii** [comment]
##  destination-bigquery-denormalized test report (commit [`face39e39e`](https://github.com/airbytehq/airbyte/pull/28721/commits/face39e39e1e5e67ede58c620f4e1a85ce32cf25)) - ✅

⏲️  Total pipeline duration: 18mn53s 

| Step                                                                                         | Result   |
|:---------------------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-bigquery-denormalized/metadata.yaml     | ✅        |
| Connector version semver check                                                               | ✅        |
| QA checks                                                                                    | ✅        |
| Build connector tar                                                                          | ✅        |
| Build destination-bigquery-denormalized docker image for platform linux/x86_64               | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-bigquery-denormalized:integrationTest | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/edgao_dest_bigquery_null_order/1690412428/face39e39e1e5e67ede58c620f4e1a85ce32cf25/destination-bigquery-denormalized/1.5.3/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery-denormalized test
```

**Pull request merged**